### ISSUE TITLE
[Snapshot] Child constraints are not included in snapshot when constraining a .value element

### ISSUE THREAD

**Issue by ewoutkramer** [ewoutkramer/fhir-net-api#827]
To reproduce:  take a look at buildOrganizationWithRegexConstraintOnType() in TestProfileArtifactSource.cs. This creates an Organization with a constraint on Organization.name.value (hence on string), adding a regex constraint (in the form of an extension) The relevant part of the differential looks like this:

```
 <element>
            <path value="Organization.name.value"/>
            <type>
                <extension url="http://hl7.org/fhir/StructureDefinition/regex">
                    <valueString value="[A-Z].*"/>
                </extension>
                <code value="string"/>
            </type>
  </element>
```

In the base (string.value), the `<type>` node looks like this:

```
<type>
     <extension url="http://hl7.org/fhir/StructureDefinition/regex">
              <valueString value="[^\s]+(\s[^\s]+)*" />
     </extension>
      <code>
                <extension url="http://hl7.org/fhir/StructureDefinition/structuredefinition-json-type">
                  <valueString value="string" />
                </extension>
                <extension url="http://hl7.org/fhir/StructureDefinition/structuredefinition-xml-type">
                  <valueString value="xsd:token" />
                </extension>
                <extension url="http://hl7.org/fhir/StructureDefinition/structuredefinition-rdf-type">
                  <valueString value="xsd:token" />
                </extension>
      </code>
</type>
```

The resulting snapshot after processing the differential shown above is:

```
            <type>
                <extension url="http://hl7.org/fhir/StructureDefinition/regex">
                    <valueString value="[A-Z].*"/>
                </extension>
                <code value="string"/>
            </type>
```

This results in loss of an important set of constraints (all encoded as extensions) in the snapshot.

The reason we have not found this before is because not many people actually do constraints on .value elements (the only usecase being adding an regex constraint).  And since the new regex constraint (which is normally more precise than the base one) IS present, all seems to work fine.

However, in R4, the validator is dependent on the additional extension on the `<type>/<code>` node and will fail to recognize constrained .value elements as primitives.  There is currently a work-around in place for this bug, which should be removed when this gets fixed (look for  `public static bool IsPrimitiveConstraint(this ElementDefinition ed) ` to find it).

**Comment by wmrutten** [comment]
Confirmed, I can reproduce the behavior. 

This line is too blame:
https://github.com/ewoutkramer/fhir-net-api/blob/afbbc28590d028bbb5c66a3e412d1dff17534bf5/src/Hl7.Fhir.Specification/Specification/Snapshot/ElementDefnMerger.cs#L121

Currently, if a differential specifies _any_ constraint on ElementDefinition.type, then this completely overrides / replaces the original type definition in the snapshot. Specifically, any extensions that may exist in the snapshot are lost.

Looks like this logic needs to be a bit smarter:
* If Element.type in the differential is empty, then inherit all types from snapshot
* If Element.type in the differential is not empty, then construct Element.type in snapshot by merging all the existing diff types on top of the associated base types. All base types w/o a matching diff type are excluded from the snapshot (i.e. these types are no longer valid in the derived profile).

**Comment by wmrutten** [comment]
Problem:

* `ElementDefnMerger.mergeComplexAttribute` calls IDeepCopyable.CopyTo()
* `Element.CopyTo()` implementation overwrites (replaces) all existing extensions in target element

Can we change the behavior of CopyTo w/o breaking compatibility?

**Comment by wmrutten** [comment]
Let's discuss this before merging.

**Comment by wmrutten** [comment]
Notes:

* `ElementDefnMerger` now implements custom merging logic specifically for `ElementDefinition.TypeRefComponent`, inheriting any extensions present on the snapshot type component (specifically "compiler magic" extensions on .value element of primitive profiles). This custom logic currently relies on hard-coded knowledge about `TypeRefComponent`.
* However for all other complex collection properties (mappings, constraints), diff collection items completely overwrite/replace snap collection items; this implies that such diff components do NOT inherit custom extensions defined by associated snap components.
* To solve this generically, all relevant complex types in the API (and all nested/referenced types...) should also implement an internal `MergeTo` method (cf. `CopyTo`) that merges diff constraints onto snapshot constraints, according to the snapshot inheritance rules defined by FHIR for the specific complex type.

`CopyTo`:

- recursively iterate matching child elements on source (diff) & target (snap)
- for each element pair, source value (if present) replaces/overwrites target value

`MergeTo`:

- recursively iterate matching child elements on source (diff) & target (snap)
- for each element pair:
  - scalar: source value (if present) replaces/overwrites target value
  - list:
    - copy all existing collection items from source to target collection
    - iterate items in diff collection
    - for each item in diff collection, find matching item in snap collection
      (Requires specific matching rules per collection; e.g. match mappings on Id property)
    - if a matching item exists in snapshot, then merge diff item onto matching snap item
    - otherwise, if no match, then add diff item to target collection

**Comment by ewoutkramer** [comment]
Fixed by #827.
